### PR TITLE
Update next of kin to an optional field

### PR DESCRIPTION
**Pull request by Rishit02**
**Inner workings**
- The Next of Kin field is now optional
- The Next of Kin attribute (in the Person class) is now expressed as an Optional object 

**Storage**
- The Next of Kin attribute is serialized and stored as a JSON object
- The Next of Kin attribute is deserialized and displayed if the Optional is not empty

**Test cases**
- Test cases are yet to be updated

**Display**
- The Display is updated, ie, visibility is toggled on if Optional is not empty, otherwise the Next of Kin label is not visible

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/UpdateCommand.java [hunk 1]
```diff
@@ -119,7 +119,8 @@ private static Person createUpdatedPerson(Person personToUpdate, UpdatePersonDes
         Address updatedAddress = updatePersonDescriptor.getAddress().orElse(personToUpdate.getAddress());
         Description updatedDescription = updatePersonDescriptor.getDescription().orElse(personToUpdate
                 .getDescription());
-        NextOfKin updatedNextOfKin = updatePersonDescriptor.getNextOfKin().orElse(personToUpdate.getNextOfKin());
+        Optional<NextOfKin> updatedNextOfKin = Optional.ofNullable(updatePersonDescriptor.getNextOfKin()
```
Comment (pjanthony2001) [inline src/main/java/seedu/address/logic/commands/UpdateCommand.java:122]: You can simplify this to `Optional<NextOfKin> updatedNextOfKin = updatePersonDescriptor.getNextOfKin().or(() -> personToUpdate.getNextOfKin())`

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 2]
```diff
@@ -39,22 +40,28 @@ public AddCommand parse(String args) throws ParseException {
...
         Phone phone = ParserUtil.parsePhone(argMultimap.getValue(PREFIX_PHONE).get());
         Email email = ParserUtil.parseEmail(argMultimap.getValue(PREFIX_EMAIL).get());
         Address address = ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS).get());
         Description description = ParserUtil.parseDescription(argMultimap.getValue(PREFIX_DESCRIPTION).get());
-        NextOfKin nextOfKin = ParserUtil.parseNextOfKin(argMultimap.getValue(PREFIX_NOK).get());
         Set<Tag> tagList = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
 
+
+        Optional<NextOfKin> nextOfKin = Optional.ofNullable(null);
+        if (argMultimap.getValue(PREFIX_NOK).isPresent()) {
```
Comment (pjanthony2001) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:61]: use `map`, so you can get rid of the if statement and just do NextOfKin nextOfKin = argMultimap.getValue(PREFIX_NOK).map(ParserUtil::parseNextOfKin);

File: src/main/java/seedu/address/logic/parser/UpdateCommandParser.java [hunk 3]
```diff
@@ -68,7 +68,8 @@ public UpdateCommand parse(String args) throws ParseException {
                     ParserUtil.parseDescription(argMultimap.getValue(PREFIX_DESCRIPTION).get()));
         }
         if (argMultimap.getValue(PREFIX_NOK).isPresent()) {
-            updatePersonDescriptor.setNextOfKin(ParserUtil.parseNextOfKin(argMultimap.getValue(PREFIX_NOK).get()));
```
Comment (pjanthony2001) [inline src/main/java/seedu/address/logic/parser/UpdateCommandParser.java:71]: Same as above, the if statement can be removed

File: src/main/java/seedu/address/model/person/NextOfKin.java [hunk 4]
```diff
@@ -9,8 +9,7 @@
  */
 public class NextOfKin {
     public static final String MESSAGE_CONSTRAINTS =
-            "Names should only contain alphanumeric characters and spaces, and it should not be blank";
-
+            "Next of Kin should only contain alphanumeric characters and spaces, and it should not be blank";
```
Comment (pjanthony2001) [inline src/main/java/seedu/address/model/person/NextOfKin.java:12]: Nice catch

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 5]
```diff
@@ -57,7 +57,15 @@ public PersonCard(Person person, int displayedIndex) {
...
         email.setText(person.getEmail().value);
         description.setText(person.getDescription().value);
-        nextOfKin.setText(person.getNextOfKin().value);
+
+        // Set text for nextOfKin label if value is present, otherwise make it invisible
+        person.getNextOfKin().ifPresentOrElse(
+                nok -> {
+                    nextOfKin.setText(nok.value);
+                    nextOfKin.setVisible(true);
+                }, () -> nextOfKin.setVisible(false)
```
Comment (pjanthony2001) [inline src/main/java/seedu/address/ui/PersonCard.java:66]: need to set not only visible but also nextOfKin.managed(false) as well. Try abstracting it out to a method,

hide(Node node) {
return () -> node.setVisible(false);
                   node.setManaged(false);
}

Also you would need setManaged(true) for the consumer then

**Review by pjanthony2001 — CHANGES_REQUESTED** [review]
The serialisation of the JSON objects might need to change, I don't think we should be calling our own `ObjectMapper` but we should be using an Annotation.

**Comment by pjanthony2001** [comment]
I sent you a file on Telegram, detailing how it is supposed to be implemented, pls take a look aye. The issue with us adding our own ObjectMapper is that there might be more bugs introduced, and if whats in the file is followed, then it would very much simplify the toModelType method.

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java (hunk 2 above)
Comment (Rishit02) [reply src/main/java/seedu/address/logic/parser/AddCommandParser.java:61]: If I use map like that I have to handle some parse exception.
Tried with ifPresent and map and having the same issue
image

**Pull request closed without merging**